Introduced support for System wide Channel Types that can be used in any

kolkoo opened the pull request:
binding. There are default system channel types and every binding can
define its own through XML.

Signed-off-by: IVAN GEORGIEV ILIEV <ivan.iliev@musala.com>

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/groovy/org/eclipse/smarthome/core/thing/xml/test/SystemWideChannelTypesTest.groovy:
```diff
@@ -0,0 +1,128 @@
+import org.junit.Test
+import org.osgi.framework.Bundle
+
+class SystemWideChannelTypesTest extends OSGiTest {
```
Could please add javadoc with@author here.

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/groovy/org/eclipse/smarthome/core/thing/xml/test/SystemWideChannelTypesTest.groovy:
```diff
@@ -0,0 +1,128 @@
+
+        def thingTypes = thingTypeProvider.getThingTypes(null).findAll{it.getUID().getId().equals("wireless-router") }
+
+        for(ThingType type : thingTypes) {
```
If for some reason the list of thing types is empty, the test will pass, but I think it should fail. Better use thingTypes.find() for example and make explicit checks, which are not within a loop

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/groovy/org/eclipse/smarthome/core/thing/xml/test/SystemWideChannelTypesTest.groovy:
```diff
@@ -0,0 +1,128 @@
+                }
+            }
+
+            assertThat i, is(0)
```
At first sight it is not clear what is really checked here. Taking also the previous comment in consideration please rewrite this to search for the channel types and check if there are present and also contain the defined properties.

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ChannelTypeConverter.java:
```diff
@@ -107,8 +109,12 @@ private StateDescription readStateDescription(NodeIterator nodeIterator) {
+        boolean advanced = readBoolean(attributes, "advanced", false);
+        boolean system = readBoolean(attributes, "system", false);
+
+        String uid = system ? String.format("%s:%s", SystemChannelTypeProvider.NAMESPACE, super.getID(attributes))
```
Could you extract the creation of the system channel type UID into a separate method please: so that this code looks like this: `String uid = system ? getSystemChannelTypeId(super.getID(attributes)) : super.getID(attributes))`

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ThingTypeXmlProvider.java:
```diff
@@ -43,9 +45,10 @@
  * process has been finished, the cache is cleared again. The merge process is started when {@link #addingFinished()} is
  * invoked from the according {@link XmlDocumentBundleTracker}.
- *
+ * 
```
You added a trailing whitespace here and I think @maggu2810 does not like this ;). Could you please apply save actions, to remove trailing white spaces? I think it also affects a lot other files.

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/ThingTypeXmlResult.java:
```diff
@@ -75,6 +77,15 @@ public ConfigDescription getConfigDescription() {
+                    int systemPrefixIdx = typeId.indexOf(SystemChannelTypeProvider.NAMESPACE_PREFIX);
+                    if (systemPrefixIdx != -1) {
+                        typeUID = String
+                                .format("%s:%s",
```
Maybe you could reuse the method to construct a system channel type I mentioned in a previous comment here.

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/XmlThingTypeProvider.java:
```diff
@@ -241,7 +253,7 @@ public ThingType getThingType(ThingTypeUID thingTypeUID, Locale locale) {
      * <p>
      * This method returns silently if the module is {@code null}.
-     *
+     * 
```
Another example for trailing white space.

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing.xml/src/main/java/org/eclipse/smarthome/core/thing/xml/internal/XmlThingTypeProvider.java:
```diff
@@ -73,9 +83,11 @@ public XmlThingTypeProvider() {
+     * 
+     * @param bundle
+     *            the module to which the Thing type to be added
+     * @param thingType
```
Did the original ESH formatter change this? Then the question is again, why it was not formatted correctly before.

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing.xml/thing-description-1.0.0.xsd:
```diff
@@ -85,9 +86,15 @@
         </xs:sequence>
     </xs:complexType>
 
+	<xs:simpleType name="namespaceIdRestrictionPattern">
```
You used tabs here. Could you please change it to white spaces.

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing/OSGI-INF/SystemChannelTypeProvider.xml:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
the file should be DefaultSystemChannelTypeProvider.xml as it registers the OSGi service for the DefaultSystemChannelTypeProvider.

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/DefaultSystemChannelTypeProvider.java:
```diff
@@ -0,0 +1,54 @@
+
+/**
+ * Implementation providing default system wide channel types and mechanisms for
+ * registration/unregistration.
```
What do you mean with "and mechanisms for registration/unregistration."?

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/DefaultSystemChannelTypeProvider.java:
```diff
@@ -0,0 +1,54 @@
+     * Low battery default system wide {@link ChannelType}
+     */
+    private static final ChannelType SYSTEM_CHANNEL_LOW_BATTERY = new ChannelType(new ChannelTypeUID(
+            "system:low-battery"), false, "Boolean", "Low Battery", null, "Battery", new HashSet<String>(
```
Item type must be "Switch" here, not "Boolean"

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/DefaultSystemChannelTypeProvider.java:
```diff
@@ -0,0 +1,54 @@
+     * Signal strength default system wide {@link ChannelType}
+     */
+    private static final ChannelType SYSTEM_CHANNEL_SIGNAL_STRENGTH = new ChannelType(new ChannelTypeUID(
+            "system:signal-strength"), false, "Number", "Signal Strength", null, "QualityOfService",
```
Could you please extract the  "QualityOfService" string into a constant name TAG_QUALITY_OF_SERVICE.

kaikreuzer replied:
Do we want to define tags at all? SO far we did not decide on a default tag library and how to proceed with tags in general. So I would rather not set any tags at the moment.

kolkoo replied:
I will remove the tags if necessary

dnobel replied:
The idea was to simply filter out such Items by a tag, which are more QoS related rather can covering functional aspects. But I agree that we first should discuss it again. So @kolkoo could you please remove it for this PR.

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/DefaultSystemChannelTypeProvider.java:
```diff
@@ -0,0 +1,54 @@
+
+    /**
+     * Signal strength default system wide {@link ChannelType}
+     */
```
Please add a short description about the channel type itself (ask also Alex).

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/DefaultSystemChannelTypeProvider.java:
```diff
@@ -0,0 +1,54 @@
+
+    /**
+     * Low battery default system wide {@link ChannelType}
+     */
```
Please add a short description about the channel type itself (ask also Alex).

dnobel commented:
The concept has one limitation, we should just be aware of. There are now 2 ways how to provide additional SystemChannelTypes:

1) By just adding an according xml file into a bundle
2) By implementing an own SystemChannelTypeProvider

So the approach is very flexible. But both ways require the bundle, which contains the additional system channel types (either the XML or the OSGi service) be loaded before the binding, which refers to the system channel type. We discussed how to solve it and came to the conclusion that it is almost impossible. 

One possibility is to apply an according start oder to the bundle, the other is make the decency explicit, by for example importing a class from the bundle, which provides the additional channel types.

For the default system channel types it should not a problem, because the DefaultSystemChannelTypes are registered within the "core.thing" bundle and no binding could be started without this bundle. So implicitly all default system channel types will be available.

dnobel commented on bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/SystemChannels.bundle/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,36 @@
+    </thing-type> 
+    
+    <!-- System Channel Type -->
+	<channel-type id="my-channel" system="true">
```
One question regarding this XML: Is it possible to define a system wide channel types via XML without defining a thing type?

kolkoo replied:
Yes, it is possible. I also added a unit test for this.
You still need to define descriptions within a <thing:thing-descriptions></thing:thing-descriptions> tag inside of a xml in ESH-INF/thing/ subfolder in order for this to work.

dnobel replied:
OK perfect.

dnobel commented:
Finished review.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/SystemChannels.bundle/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,10 @@
+Bundle-Name: Synthetic Test Bundle
+Bundle-SymbolicName: SystemChannels.bundle;singleton:=true
+Bundle-Vendor: Eclipse.org/SmartHome
+Bundle-Version: 0.7.0.qualifier
```
we should probably not try to align this version with the code base (then it should anyhow be 0.8.0). So maybe better go for something like 1.0.0 or 0.0.1?

kolkoo replied:
we can do so - I've copied this from the other synthetic bundles in this .test bundle so they will also need to be edited

dnobel replied:
Yes, I think it was initially done by mistake, also because a MANIFEST.MF from an original ESH bundle was copied. We can use any version there, because the tests don´t care about the version. So maybe you can also change it in the other manifest files.

kolkoo replied:
So which version should I use after all?

dnobel replied:
1.0.0

dnobel commented:
The PR would be fine for me, if we can live with the limitation I mentioned above. I will leave the merge up to you.

kolkoo commented:
Could you wait for my go ahead to merge this, I want to test it out on a homebase with contributed system specific channels.

eclipsewebmaster commented:
:white_check_mark: 
#### Eclipse validation results:
All committers passed Eclipse CLA and Signed-off-by validation.

eclipsewebmaster commented:
:white_check_mark: 
#### Eclipse validation results:
All committers passed Eclipse CLA and Signed-off-by validation.

kolkoo commented:
Ok I've fixed some problems with the tests and verified everything on a homebase.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/DefaultSystemChannelTypeProvider.java:
```diff
@@ -0,0 +1,52 @@
+     * Signal strength default system wide {@link ChannelType}. Represents signal strength of a device as a number
+     * with values 0, 1, 2, 3 or 4, 0 being worst strength and 4 being best strength.
+     */
+    private static final ChannelType SYSTEM_CHANNEL_SIGNAL_STRENGTH = new ChannelType(new ChannelTypeUID(
```
Wouldn't it be useful to make these constants public (and thus move them to a constants class in an exported package)? This would allow ThingTypeProviders to refer to them easily.

dnobel replied:
Do you mean putting the ChannelTypeUID into an exported constants class? I think that makes sense, especially for dynamic ThingTypeProviders. But we should not move the whole ChannelTypeDefinition into the constants, because they can be retrieved via the getSystemChannelTypes() method.

kaikreuzer replied:
> Do you mean putting the ChannelTypeUID into an exported constants class?
This is what I though first as well. But looking into the code of a ThingTypeProvider, it has to provide ```ThingType```s, which have ```ChannelDefinition```s, which in turn have ```ChannelType```s and not ```ChannelTypeUID```s - so that's why I think the whole ChannelType should be made available as a constant. A ThingTypeProvider could then easily use it in the constructor of a ChannelDefinition. Having to get hold of it by processing a list of types returned by an OSGi service is imho no option.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/type/SystemChannelTypeProvider.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.Collection;
+
+/**
+ * A service responsible for providing System Wide {@link ChannelType}s.
```
"systemwide" instead of "System Wide"

kaikreuzer commented:
Thanks Ivan! Could you please also apply Dennis' remarks regarding the formatter (e.g. trailing spaces)?

One question about the PR from me: How is i18n dealt with? Could you maybe add an example and a test case that comes with entries in the i18n folder?

And talking about examples: What about a short section in the documentation about this new feature?

eclipsewebmaster commented:
:white_check_mark: 
#### Eclipse validation results:
All committers passed Eclipse CLA and Signed-off-by validation.

dnobel commented:
> Thanks Ivan! Could you please also apply Dennis' remarks regarding the formatter (e.g. trailing spaces)?

Ivan told me, that if he applies the formatter as a save action, it automatically adds the white spaces to all files. There is another save action for removing trailing white spaces. So we both were confused at the end.

eclipsewebmaster commented:
:white_check_mark: 
#### Eclipse validation results:
All committers passed Eclipse CLA and Signed-off-by validation.

kolkoo commented:
I18n is working the same way for system channel types as for others, when we obtain a locale specific ThingType using system channels we also obtain a locale specific ChannelType object for them, the implementation will then be looking for the following properties:
channel-type.system.<channel id>.<property> = <local specific text>
This means we can add bundle specific localization for any system channel type.
I've added a test for this.

I've fixed the system wide to -> systemwide, made the default ChannelTypes static declarations public and moved the DefaultSystemChannelTypeProvider that contains them to an exported package. Also updated the documentation.

eclipsewebmaster commented:
:white_check_mark: 
#### Eclipse validation results:
All committers passed Eclipse CLA and Signed-off-by validation.

kaikreuzer commented:
Thanks Ivan, looks great!
